### PR TITLE
consul: support WithServiceResolver option

### PR DESCRIPTION
**Pull request by Cluas**
#### Description (what this PR does / why we need it):
add endpoints resolver option

#### Which issue(s) this PR fixes (resolves / be part of):

fixes #

### CODE REVIEW

File: contrib/registry/consul/registry.go [hunk 1]
```diff
@@ -26,6 +26,15 @@ func WithHealthCheck(enable bool) Option {
 	}
 }
 
+// WithEndpointsResolver with endpoint function option.
+func WithEndpointsResolver(fn Resolver) Option {
```
Comment (tonybase) [inline contrib/registry/consul/registry.go:30]: WithEndpointResolver

File: contrib/registry/consul/client.go [hunk 2]
```diff
@@ -18,15 +18,32 @@ type Client struct {
...
 
 // NewClient creates consul client
 func NewClient(cli *api.Client) *Client {
-	c := &Client{cli: cli}
+	c := &Client{cli: cli, resolver: defaultResolver}
 	c.ctx, c.cancel = context.WithCancel(context.Background())
 	return c
 }
 
+func defaultResolver(_ context.Context, entry *api.ServiceEntry) []string {
```
Comment (tonybase) [inline contrib/registry/consul/client.go:33]: `[]*api.ServiceEntry -> []*registry.ServiceInstance`

应该是直接可以完整定制这个转换的，只是我们给了个默认的实现

Comment (Cluas) [reply contrib/registry/consul/client.go:33]: 那上面是不是不应该叫EndpointResolver 应该叫ServiceResolver

File: contrib/registry/consul/registry.go (hunk 1 above)
Comment (Cluas) [reply contrib/registry/consul/registry.go]: > WithEndpointResolver

WithServiceResolver

File: contrib/registry/consul/client.go [hunk 3]
```diff
@@ -64,11 +55,29 @@ func (d *Client) Service(ctx context.Context, service string, index uint64, pass
 			Endpoints: endpoints,
 		})
 	}
-	return services, meta.LastIndex, nil
+
+	return services
+}
+
+// Resolver is used to resolve service endpoints
+type Resolver func(ctx context.Context, entries []*api.ServiceEntry) []*registry.ServiceInstance
```
Comment (Cluas) [inline contrib/registry/consul/client.go:63]: @tonybase 这里是否命名统一一下 ServiceResolver?

**Review by tonybase — APPROVED** [review]

**Pull request merged**